### PR TITLE
Add __torch_function__ support for torch.nn.functional.embedding

### LINKED ISSUE #54292: torch.nn.functional.embedding does not participate in `__torch_function__` dispatch
Unlike other functionals, `embedding` does not have code to handle `__torch_function__` dispatch: https://github.com/pytorch/pytorch/blob/bbb06c05a833e168b10356a0c103dea7beaa3eca/torch/nn/functional.py#L2004

cc @hameerabbasi @rgommers @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54478 Add __torch_function__ support for torch.nn.functional.embedding**

Fixes #54292

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27264179](https://our.internmc.facebook.com/intern/diff/D27264179)

### CODE REVIEW

**Comment by ezyang** [comment]
oddly torch.nn.functional.embedding is already in the testing list and I verified that the new branch gets hit, so there is probably something wrong with the tests where they fail to notice if you don't hit the correct override.

**Comment by rgommers** [comment]
> oddly torch.nn.functional.embedding is already in the testing list and I verified that the new branch gets hit, so there is probably something wrong with the tests where they fail to notice if you don't hit the correct override

@hameerabbasi could you look into this test issue?

**Comment by hameerabbasi** [comment]
These are the lines that are supposed to check if the underlying function was the one passed into `__torch_function__`:

https://github.com/pytorch/pytorch/blob/8518b0ee55a3b04353f9ea3b39d142bf10359bf4/test/test_overrides.py#L624-L625

What I'm assuming happened here is that it got set to `True` in an earlier test because it's a global, and never got reset. I'll write a context manager that resets it.

**Comment by hameerabbasi** [comment]
Looking into this in #54492.

**Comment by hameerabbasi** [comment]
There are a number of other `torch.nn.functional` functions that aren't directly covered:

```
FAILED test/test_overrides.py::TestTorchFunctionOverride::test_torch_nn_functional_bilinear - AssertionError: False is not true
FAILED test/test_overrides.py::TestTorchFunctionOverride::test_torch_nn_functional_ctc_loss - AssertionError: False is not true
FAILED test/test_overrides.py::TestTorchFunctionOverride::test_torch_nn_functional_embedding - AssertionError: False is not true
FAILED test/test_overrides.py::TestTorchFunctionOverride::test_torch_nn_functional_pairwise_distance - AssertionError: False is not true
```

(from #54492)

File: torch/nn/functional.py [hunk 1]
```diff
@@ -2001,6 +2001,9 @@ def embedding(
                  [ 0.6262,  0.2438,  0.7471]]])
     """
 
+    if has_torch_function_variadic(input, weight):
```
Comment (jbschlosser) [inline torch/nn/functional.py:2004]: Could fix the lint check by formatting this to match`F.embedding_bag`:
```python
if has_torch_function_variadic(input, weight):
    return handle_torch_function(
        embedding,
        (input, weight),
        input,
        weight,
        padding_idx=padding_idx,
        max_norm=max_norm,
        norm_type=norm_type,
        scale_grad_by_freq=scale_grad_by_freq,
        sparse=sparse,
    )
```

**Comment by ezyang** [comment]
Please don't hold up approves on lint errors :) (now fixed)

**Review by jbschlosser — APPROVED** [review]

**Review by jamesr66a — APPROVED** [review]
Thanks so much!

**Pull request closed without merging**